keps/969-GangScheduling/README.md

```diff
@@ -0,0 +1,589 @@
+	JobSetAsGang GangPolicyOptions = "JobSetAsGang"
+
+	// JobSetGangPerReplicatedJob means that each ReplicatedJob will be a separate gang.
+	// Each ReplicatedJob's pods will be gang-scheduled independently.
+	JobSetGangPerReplicatedJob GangPolicyOptions = "JobSetGangPerReplicatedJob"
+
+	// JobSetWorkloadTemplate means that the JobSet will create the Workload from the provided template.
+	// This is an advanced option that provides fine-grained control over pod groups
+	// and scheduling logic for a JobSet.
+	JobSetWorkloadTemplate GangPolicyOptions = "JobSetWorkloadTemplate"
+)
+
+type GangPolicy struct {
+	// gangPolicyOption determines the gang scheduling policy for JobSet.
+	// Defaults to nil (no gang scheduling).
+	// +kubebuilder:validation:Enum=JobSetAsGang;JobSetGangPerReplicatedJob;JobSetWorkloadTemplate
+	// +optional
+	GangPolicyOption *GangPolicyOptions `json:"gangPolicyOption,omitempty"`
+
+	// workload provides a Workload template to create on JobSet creation.
+	// This field is only valid when gangPolicyOption is JobSetWorkloadTemplate.
+	// When specified, the JobSet controller will create this Workload object
+	// and link it to the JobSet for gang scheduling.
+	// +optional
+	Workload *schedulingv1alpha1.Workload `json:"workload,omitempty"`
+}
+
+// In JobSetSpec:
+type JobSetSpec struct {
+	// ... existing fields ...
+
+	// gangPolicy configures gang scheduling policy for this JobSet.
+	// Gang scheduling ensures that all pods (or groups of pods) are scheduled
+	// atomically, preventing resource deadlock in distributed workloads.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
+	// +optional
+	GangPolicy *GangPolicy `json:"gangPolicy,omitempty"`
+
+	// ... existing fields ...
+}
+```
+
+### Implementation
+
+The JobSet controller will implement gang scheduling as follows:
+
+1. **Workload Creation**:
+   - When a JobSet with `gangPolicy` is created, the controller creates a corresponding Workload object
+   - For `JobSetAsGang`: Create a single Workload with one PodSet containing all pods
+   - For `JobSetGangPerReplicatedJob`: Create a single Workload with multiple PodSets (one per ReplicatedJob)
+   - For `JobSetWorkloadTemplate`: Create the user-provided Workload template
+
+2. **Ownership and Lifecycle**:
+   - The Workload is owned by the JobSet (via OwnerReference) and shares its lifecycle
+   - Deleting the JobSet cascades deletion to the Workload
+   - The Workload name follows the pattern: `<jobset-name>-workload`
+
+3. **Job Creation Coordination**:
+   - Jobs are created with appropriate labels/annotations to link them to the Workload's PodSets
+   - The scheduler (e.g., Kueue) uses these labels to enforce gang scheduling constraints
+
+4. **Status Synchronization**:
+   - JobSet status is updated based on both Job status and Workload admission status
+   - New condition: `WorkloadAdmitted` indicates when the gang has been admitted for scheduling
```

Comment:
How would we check that gang has been admitted?

Reply:
I need to update this actually.

Right now there is no status field on Workloads and I think the only way to tell if a gang has been admitted is if the workload is scheduled.

I believe there is a KEP to add more status fields to Workload but it is not yet here. I don't see conditions or events. So the way may be (for now) to read the kubescheduler logs.